File: 03_divide_and_conquer/points_and_segments/points_and_segments.cpp
#include <iostream>
#include <vector>
#include<algorithm>

using std::vector;

bool srt(const vector<int> & v1, const vector<int> & v2){
	return v1[0]<v2[0];
}
vector<int> fast_count_segments(vector<int> starts, vector<int> ends, vector<int> points) {
  vector<int> cnt(points.size());
  //write your code here
  vector<vector <int> > seg(starts.size(),vector<int>(2,0));
  for(int i=0;i<starts.size();i++){
	seg[i][0]=starts[i];
	seg[i][1]=ends[i];}
  sort(seg.begin(),seg.end(),srt);
  for (size_t i = 0; i < points.size(); i++) {
    for (size_t j = 0; j < seg.size(); j++) {
      cnt[i] += seg[j][0] <= points[i] && points[i] <= seg[j][1];
	if(points[i]<seg[j][0])
		break;
    }
  }
  return cnt;
}

vector<int> naive_count_segments(vector<int> starts, vector<int> ends, vector<int> points) {
  vector<int> cnt(points.size());
  for (size_t i = 0; i < points.size(); i++) {
    for (size_t j = 0; j < starts.size(); j++) {
      cnt[i] += starts[j] <= points[i] && points[i] <= ends[j];
    }
  }
  return cnt;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<int> starts(n), ends(n);
  for (size_t i = 0; i < starts.size(); i++) {
    std::cin >> starts[i] >> ends[i];
  }
  vector<int> points(m);
  for (size_t i = 0; i < points.size(); i++) {
    std::cin >> points[i];
  }
  //use fast_count_segments
  vector<int> cnt = fast_count_segments(starts, ends, points);
  for (size_t i = 0; i < cnt.size(); i++) {
    std::cout << cnt[i] << ' ';
  }
}

File: 03_divide_and_conquer/points_and_segments/points_and_segments.py
# Task. You are given a set of points on a line and a set of segments on a line. The goal is to compute, for each point, the number of segments that contain this point.
# Uses python3
import sys

def fast_count_segments(starts, ends, points):
    cnt = [0] * len(points)
    seg= [list(x) for x in zip(starts, ends)]  
    seg.sort(reverse=False)
    for i in range(len(points)):
        for j in rane(len(starts)):
            cnt[i] += seg[j][0]<=points[i] and points[i]<=seg[j][1]
            if points[i]<seg[j][0]:
                break
    return cnt

def naive_count_segments(starts, ends, points):
    cnt = [0] * len(points)
    for i in range(len(points)):
        for j in range(len(starts)):
            if starts[j] <= points[i] <= ends[j]:
                cnt[i] += 1
    return cnt

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    m = data[1]
    starts = data[2:2 * n + 2:2]
    ends   = data[3:2 * n + 2:2]
    points = data[2 * n + 2:]
    #use fast_count_segments
    cnt = naive_count_segments(starts, ends, points)
    for x in cnt:
        print(x, end=' ')

